core/trino-main/src/main/java/io/trino/json/ir/IrAccessor.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.json.ir;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import io.trino.spi.type.Type;
+
+import java.util.Objects;
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+public abstract class IrAccessor
+        extends IrPathNode
+{
+    protected final IrPathNode base;
+
+    IrAccessor(IrPathNode base, Optional<Type> type)
+    {
+        super(type);
+        this.base = requireNonNull(base, "accessor base is null");
+    }
+
+    @Override
+    protected <R, C> R accept(IrJsonPathVisitor<R, C> visitor, C context)
+    {
+        return visitor.visitIrAccessor(this, context);
+    }
+
+    @JsonProperty
+    public IrPathNode getBase()
+    {
+        return base;
+    }
+
+    @Override
+    public boolean equals(Object obj)
+    {
+        if (this == obj) {
+            return true;
+        }
+        if (obj == null || getClass() != obj.getClass()) {
+            return false;
+        }
+        IrAccessor other = (IrAccessor) obj;
+        return Objects.equals(this.base, other.base);
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return Objects.hash(base);
+    }
```

Comment:
don't you need to look at super.type here?

Reply:
(and elswhere)

Reply:
`type` is intentionally skipped in `equals()` and `hashCode()` methods of all `IrPathNode`s. I wanted to compare just the structure. Type is a function of the other properties, and it might be optionally filled or not, depending on when and how the node is created - e.g. either initially or by some optimization that will be added in the future (like constant folding, tree flattening).

Reply:
>Also why can't we always have type set?

We might not know the type initially, and learn it later, for example when a JSON parameter appears to be a constant literal. Also, some nodes, like `keyValueMethod()`, which returns JSON objects, inherently cannot have the SQL type.

Generally, recording the types serves two purposes: 
- to statically validate the JSON path,
- to enable optimizations, like pre-resolving operators for known types.

Reply:
>Hmm - worth a comment - not sure what is the good place for that though; probably the best one is in IrPathNode where type is declared.

comment added